xtable-hive-metastore/src/main/java/org/apache/xtable/hms/HMSCatalogPartitionSyncOperations.java

```diff
@@ -0,0 +1,222 @@
+import org.apache.hadoop.hive.metastore.api.Table;
+import org.apache.thrift.TException;
+
+import org.apache.hudi.common.util.CollectionUtils;
+import org.apache.hudi.exception.TableNotFoundException;
+
+import org.apache.xtable.catalog.CatalogPartition;
+import org.apache.xtable.catalog.CatalogPartitionSyncOperations;
+import org.apache.xtable.exception.CatalogSyncException;
+import org.apache.xtable.model.catalog.CatalogTableIdentifier;
+import org.apache.xtable.model.catalog.HierarchicalTableIdentifier;
+
+@Log4j2
+public class HMSCatalogPartitionSyncOperations implements CatalogPartitionSyncOperations {
+
+  private final IMetaStoreClient metaStoreClient;
+  private final HMSCatalogConfig catalogConfig;
+
+  public HMSCatalogPartitionSyncOperations(
+      IMetaStoreClient metaStoreClient, HMSCatalogConfig hmsCatalogConfig) {
+    this.metaStoreClient = metaStoreClient;
+    this.catalogConfig = hmsCatalogConfig;
+  }
+
+  @Override
+  public List<CatalogPartition> getAllPartitions(CatalogTableIdentifier catalogTableIdentifier) {
+    HierarchicalTableIdentifier tableIdentifier =
+        toHierarchicalTableIdentifier(catalogTableIdentifier);
+    try {
+      return metaStoreClient
+          .listPartitions(
+              tableIdentifier.getDatabaseName(), tableIdentifier.getTableName(), (short) -1)
+          .stream()
+          .map(p -> new CatalogPartition(p.getValues(), p.getSd().getLocation()))
+          .collect(Collectors.toList());
+    } catch (TException e) {
+      throw new CatalogSyncException(
+          "Failed to get all partitions for table " + tableIdentifier, e);
+    }
+  }
+
+  @Override
+  public void addPartitionsToTable(
+      CatalogTableIdentifier catalogTableIdentifier, List<CatalogPartition> partitionsToAdd) {
+    HierarchicalTableIdentifier tableIdentifier =
+        toHierarchicalTableIdentifier(catalogTableIdentifier);
+    if (partitionsToAdd.isEmpty()) {
+      log.info("No partitions to add for " + tableIdentifier);
+      return;
+    }
+    log.info("Adding partitions " + partitionsToAdd.size() + " to table " + tableIdentifier);
+    try {
+      StorageDescriptor sd =
+          metaStoreClient
+              .getTable(tableIdentifier.getDatabaseName(), tableIdentifier.getTableName())
+              .getSd();
+      for (List<CatalogPartition> batch :
+          CollectionUtils.batches(partitionsToAdd, catalogConfig.getMaxPartitionsPerRequest())) {
+        List<org.apache.hadoop.hive.metastore.api.Partition> partitionList = new ArrayList<>();
+        batch.forEach(
+            partition -> {
+              StorageDescriptor partitionSd = new StorageDescriptor();
+              partitionSd.setCols(sd.getCols());
+              partitionSd.setInputFormat(sd.getInputFormat());
+              partitionSd.setOutputFormat(sd.getOutputFormat());
+              partitionSd.setSerdeInfo(sd.getSerdeInfo());
+
+              partitionSd.setLocation(partition.getStorageLocation());
```

Comment:
There is similar code on line 134 below. Can we move some of this into a common method to reduce repetition?